pyspi/base.py

```diff
@@ -1,150 +1,122 @@
-
-        return function(self,data,i=i)
+            i = 0 if data.n_processes == 1 else ValueError("Please specify `i`.")
+        return func(self, data, i=i)
+    return wrapper
 
-    return parsed_function
-
-def parse_bivariate(function):
-    def parsed_function(self,data,data2=None,i=None,j=None,inplace=True):
-        if not isinstance(data,Data):
+def parse_bivariate(func):
+    def wrapper(self, data, data2=None, i=None, j=None, inplace=True):
+        if not isinstance(data, Data):
             if data2 is None:
-                raise TypeError('Input must be either a pyspi.data object or two 1D-array inputs.'
-                                    f' Received {type(data)} and {type(data2)}.')
-            data1 = data
+                raise TypeError("Provide either a Data object or two 1D arrays.")
             data = Data()
-            data.add_process(data1)
+            data.add_process(data)
             data.add_process(data2)
         elif not inplace:
-            # Ensure we don't write over the original
             data = copy.deepcopy(data)
-
-        if i is None and j is None:
+        if i is None or j is None:
             if data.n_processes == 2:
-                i,j = 0,1
+                i, j = 0, 1
             else:
-                Warning('i and j not set.')
-
-        return function(self,data,i=i,j=j)
-
-    return parsed_function
-
-def parse_multivariate(function):
-    def parsed_function(self,data,inplace=True):
-        if not isinstance(data,Data):
-            # Create a pyspi.Data object from iterable data object
-            try:
-                procs = data
-                data = Data()
-                for p in procs:
-                    data.add_process(p)
-            except IndexError:
-                raise TypeError('Data must be either a pyspi.data.Data object or an and iterable of numpy.ndarray''s.')
+                raise ValueError("Indices i and j must be provided.")
+        return func(self, data, i=i, j=j)
+    return wrapper
+
+def parse_multivariate(func):
+    def wrapper(self, data, inplace=True):
+        if not isinstance(data, Data):
+            data = Data()
+            for p in data:
+                data.add_process(p)
         elif not inplace:
-            # Ensure we don't write over the original
             data = copy.deepcopy(data)
-
-        return function(self,data)
-
-    return parsed_function
-
+        return func(self, data)
+    return wrapper
+
+# ------------------------------
+# Base SPI class
+# ------------------------------
+class BaseSPI(BaseObject):
+    _tags = {
+        "capability-multivariate": True,
+        "capability-bivariate": True,
+        "capability-unequal_length": False,
+        "python_dependencies": "sktime"
+    }
+
+    def _spi(self, data: Data, i: int = 0) -> float:
+        raise NotImplementedError("Subclass must implement _spi.")
+
+    @parse_univariate
```

Comment:
I would not use these as decorators at all! But move the logic as a call into `spi`.